File: revisiting_ctci/chapter3/MultiStack_3_1.go
/*
use one array to implement three stacks.

i.e.

array := []Stack{3}

have functions to call the array position
and perform the stacks command on the given position

*/

package main

import (
	"./stack"
	"fmt"
)

type MultiStack [3]stack.Stack

func (m MultiStack) String() string {
	outstring := ""
	for i, dat := range m {
		outstring = fmt.Sprintf("%v%v\t%v\n", outstring, i, dat)
	}
	return outstring
}

func (m *MultiStack) Push(dat, pos int) {
	m[pos].Push(dat)
}

func (m *MultiStack) Pop(pos int) int {
	output := m[pos].Pop()
	return output
}

func (m MultiStack) IsEmpty(pos int) bool {
	return m[pos].IsEmpty()
}

func main() {

	test := MultiStack{}
	fmt.Println(test.IsEmpty(0))
	test.Push(1, 0)
	test.Push(2, 0)
	test.Push(3, 0)
	test.Push(4, 0)
	fmt.Println(test)

	fmt.Println(test.IsEmpty(0))
	test.Pop(0)
	fmt.Println(test)

	test.Push(4, 1)
	test.Push(5, 1)
	test.Push(6, 1)

	test.Push(7, 2)
	test.Push(8, 2)
	test.Push(9, 2)

	fmt.Println(test)

}

File: revisiting_ctci/chapter3/MultiStack_3_1.py
"""
use one array to implement three stacks.

i.e.

array := []Stack{3}

have functions to call the array position
and perform the stacks command on the given position
"""


from stack import Stack


class MultiStack():
	def __init__(self):
		self.__stack_list = [Stack(), Stack(), Stack()]

		self.__pos_err = 'MultiStack has three positions avaliable!' + \
								' index: 0 -> 2.'
	def Push(self, dat, pos):
		""" take the given data and add it to the proper stack """
		if pos > 2 :
			raise IndexError(self.__pos_err)

		self.__stack_list[pos].Push(dat)

	def Pop(self, pos):
		if pos > 2 :
			raise IndexError(self.__pos_err)

		return self.__stack_list[pos].Pop()

	def Peek(self, pos):
		if pos > 2 :
			raise IndexError(self.__pos_err)

		return self.__stack_list[pos].Peek()

	def __repr__(self):
		outstring = ''
		for i, x in enumerate(self.__stack_list):
			dat = f'{i}:\t{x}\n'
			outstring += dat
		return outstring

if __name__ == '__main__':

	test = MultiStack()

	test.Push(1,0)	
	test.Push(2,0)	
	test.Push(3,0)	
	test.Push(4,0)	

	test.Push(5,1)	
	test.Push(6,1)	
	test.Push(7,1)	
	test.Push(8,1)	

	test.Push(9,2)	
	test.Push(10,2)	
	test.Push(11,2)	
	test.Push(12,2)

	print(test)
	print('Popped vals:')
	print(test.Pop(0))
	print(test.Pop(1))
	print(test.Pop(2))
	print('\n')

	print(test)